File: testdata/fixtures/swift-small/shapes.swift
// shapes defines a Shape protocol, a Base/Circle class pair, a Point struct
// conforming to Shape via an extension, and a Kind enum.

import Foundation

/// Shape is conformed to by geometric figures.
protocol Shape {
    func area() -> Double
    func label() -> String
}

/// Base is the common superclass for shapes.
class Base {
    var name: String

    init(name: String) {
        self.name = name
    }

    func label() -> String {
        return name
    }
}

/// Circle is a circle with a radius, conforming to Shape.
class Circle: Base, Shape {
    let radius: Double

    init(radius: Double) {
        self.radius = radius
        super.init(name: "circle")
    }

    func area() -> Double {
        return PI * radius * radius
    }

    override func label() -> String {
        return "circle"
    }
}

/// Point is a 2D point.
struct Point {
    var x: Double
    var y: Double
}

/// Point conforms to Shape via an extension.
extension Point: Shape {
    func area() -> Double {
        return x * y
    }

    func label() -> String {
        return "point"
    }
}

/// Kind enumerates the supported shape families.
enum Kind {
    case round
    case polygon
    case custom
}

let PI = 3.14159

File: testdata/fixtures/rs-small/shapes.rs
// shapes defines a Shape trait and a Circle struct implementing it.
pub mod shapes {
    /// Shape is implemented by geometric figures.
    pub trait Shape {
        fn area(&self) -> f64;
        fn label(&self) -> String;
    }

    /// Circle is a circle with a radius.
    pub struct Circle {
        pub radius: f64,
    }

    impl Circle {
        /// new builds a Circle of the given radius.
        pub fn new(radius: f64) -> Circle {
            Circle { radius }
        }

        pub fn diameter(&self) -> f64 {
            self.radius * 2.0
        }
    }

    impl Shape for Circle {
        fn area(&self) -> f64 {
            3.14159 * self.radius * self.radius
        }

        fn label(&self) -> String {
            String::from("circle")
        }
    }

    /// Kind enumerates the supported shape families.
    pub enum Kind {
        Round,
        Polygon,
        Custom(u8),
    }

    pub const PI: f64 = 3.14159;
}
